src/network/message.rs

```diff
@@ -160,50 +160,69 @@ pub enum NetworkMessage {
-            NetworkMessage::Version(_) => "version",
-            NetworkMessage::Verack     => "verack",
-            NetworkMessage::Addr(_)    => "addr",
-            NetworkMessage::Inv(_)     => "inv",
-            NetworkMessage::GetData(_) => "getdata",
-            NetworkMessage::NotFound(_) => "notfound",
-            NetworkMessage::GetBlocks(_) => "getblocks",
-            NetworkMessage::GetHeaders(_) => "getheaders",
-            NetworkMessage::MemPool    => "mempool",
-            NetworkMessage::Tx(_)      => "tx",
-            NetworkMessage::Block(_)   => "block",
-            NetworkMessage::Headers(_) => "headers",
-            NetworkMessage::SendHeaders => "sendheaders",
-            NetworkMessage::GetAddr    => "getaddr",
-            NetworkMessage::Ping(_)    => "ping",
-            NetworkMessage::Pong(_)    => "pong",
-            NetworkMessage::GetCFilters(_) => "getcfilters",
-            NetworkMessage::CFilter(_) => "cfilter",
-            NetworkMessage::GetCFHeaders(_) => "getcfheaders",
-            NetworkMessage::CFHeaders(_) => "cfheaders",
-            NetworkMessage::GetCFCheckpt(_) => "getcfcheckpt",
-            NetworkMessage::CFCheckpt(_) => "cfcheckpt",
-            NetworkMessage::Alert(_)    => "alert",
-            NetworkMessage::Reject(_)    => "reject",
+            NetworkMessage::Version(_) => &Self::possibilites()[0],
+            NetworkMessage::Verack => &Self::possibilites()[1],
+            NetworkMessage::Addr(_) => &Self::possibilites()[2],
+            NetworkMessage::Inv(_) => &Self::possibilites()[3],
+            NetworkMessage::GetData(_) => &Self::possibilites()[4],
+            NetworkMessage::NotFound(_) => &Self::possibilites()[5],
+            NetworkMessage::GetBlocks(_) => &Self::possibilites()[6],
+            NetworkMessage::GetHeaders(_) => &Self::possibilites()[7],
+            NetworkMessage::MemPool => &Self::possibilites()[8],
+            NetworkMessage::Tx(_) => &Self::possibilites()[9],
+            NetworkMessage::Block(_) => &Self::possibilites()[10],
+            NetworkMessage::Headers(_) => &Self::possibilites()[11],
+            NetworkMessage::SendHeaders => &Self::possibilites()[12],
+            NetworkMessage::GetAddr => &Self::possibilites()[13],
+            NetworkMessage::Ping(_) => &Self::possibilites()[14],
+            NetworkMessage::Pong(_) => &Self::possibilites()[15],
+            NetworkMessage::GetCFilters(_) => &Self::possibilites()[16],
+            NetworkMessage::CFilter(_) => &Self::possibilites()[17],
+            NetworkMessage::GetCFHeaders(_) => &Self::possibilites()[18],
+            NetworkMessage::CFHeaders(_) => &Self::possibilites()[19],
+            NetworkMessage::GetCFCheckpt(_) => &Self::possibilites()[20],
+            NetworkMessage::CFCheckpt(_) => &Self::possibilites()[21],
+            NetworkMessage::Alert(_) => &Self::possibilites()[22],
+            NetworkMessage::Reject(_) => &Self::possibilites()[23],
```

Comment:
This is quite unfortunate.
what I tried to do is make changes to enums visible at compile time, that way if someone adds an enum variant the tests won't compile without fixing them too.

I did it here using a static array that is also used in tests, but this mean it's harder to review this properly, not sure if the tradeoff is worth it or not.

Reply:
Yeah, I think it's better to inline the `possibilities` function and accept that the compiler may allow enum variants to be extended incompletely. In some Blockstream code I have used macros to accomplish this but the result isn't much prettier.

Reply:
The indices are definitely unreadable. There is no way to verify any of this with the naked eye, so any malicious code somehow skipping the code testing this could break everything.

I also don't think this is needed per se. The `CommandString` rng should allow all UTF8 strings of 8 characters. And then we can have a separate `NetworkMessage` rng that does the match trick like you did for `RejectReason` to make sure it has all options.